### PR TITLE
Tag Processor: Add bookmark invalidation logic

### PR DESCRIPTION
**Pull request by ockham**
## What?

Part of #44410.

Invalidate bookmarks in a `WP_HTML_Tag_Processor` instance if the tags that they're pointing to have been replaced.

## Why?
While `WP_HTML_Tag_Processor` currently only supports changing a given tag's attributes, we plan to also provide methods to make broader changes (e.g. [`set_content_inside_balanced_tags `](https://github.com/WordPress/gutenberg/pull/47036)) (probably through a subclass of `WP_HTML_Tag_Processor`).

An API like that will have the potential of replacing a tag that a bookmark points to. As a preparation, we thus need to make sure that all bookmarks affected by a HTML replacement are invalidated (i.e. released).

## How?
By extending the existing loop in `apply_attributes_updates` that adjusts bookmarks' start and end positions upon HTML changes to check if the entire bookmark is within a portion of the HTML that has been replaced.

Note that this PR is extracted from https://github.com/WordPress/gutenberg/pull/47068, per [this suggestion](https://github.com/WordPress/gutenberg/pull/47068#issuecomment-1405439927).

## Testing Instructions
As stated above, there aren't currently any methods that actually replace existing bookmarks, so the relevant codepath should never be hit. Furthermore, `apply_attributes_updates` is a `private` method, so we cannot directly test it.

However, we have extensive unit test coverage of `WP_HTML_Tag_Processor`'s public methods which guarantee that their desired behavior is still warranted.

Finally, https://github.com/WordPress/gutenberg/pull/47036 [contains](https://github.com/WordPress/gutenberg/pull/47036/files#diff-93724b2f6e5dbf16c2b3c0efea57fa442b572f75f325d1cb185cf614b692cf9fR1306-R1347) the code from this PR, and has [unit test coverage for bookmark invalidation](https://github.com/WordPress/gutenberg/pull/47036/files#diff-46de4f3b288b4df1647366b416c84f02b7296f8a3040123513021662c2aab15fR288-R303).

### CODE REVIEW

**Review by dmsnell — COMMENTED** [review]
is this something you consider necessary for 6.1?
is this necessary at all? It appears to change the behavior of bookmarks where if we seek to a bookmark we previously set, but that bookmark is gone, instead of returning `false`, with `WP_DEBUG` set it will crash saying "you did it wrong."

is the safety we're after something that belongs in `seek` instead so that we can still distinguish between "failed to seek to the bookmark you previously set" and "you tried to seek to a non-existent bookmark"? that is, "something failed because of normal situations" vs. "you tried to do something that is guaranteed to fail and should never be done"

**Comment by ockham** [comment]
> is this something you consider necessary for 6.1? 

You mean 6.2 🙂 No, this isn't needed for the pending WP release.

> is this necessary at all?

I kinda think it is 🤔 Since a bookmark is just a `WP_HTML_Span` -- i.e. a start/end tuple -- it could end up pointing to something totally different.

Take [this unit test](https://github.com/WordPress/gutenberg/pull/47036/files#diff-46de4f3b288b4df1647366b416c84f02b7296f8a3040123513021662c2aab15fR288-R303) for example. We start with the following HTML, and set a bookmark to the `img` tag:

```html
<div>outside</div><section><div><img>inside</div></section>
                                ^   ^
```

Now we replace the contents between the third and fourth `div` tags:

```html
<div>outside</div><section><div>This is the new section content.</section>
                                ^   ^
```

The bookmark isn't even pointing to a tag anymore. I don't think we want the user to proceed assuming that all is well with that bookmark 😬 

It's easy enough for us to add logic (like in this PR) to determine which bookmarks (if any) have been invalidated; it seems like it'd be quite cumbersome for the user to do so (without access to class internals).

> It appears to change the behavior of bookmarks where if we seek to a bookmark we previously set, but that bookmark is gone, instead of returning `false`, with `WP_DEBUG` set it will crash saying "you did it wrong."

Yeah, that's a fair point.

> is the safety we're after something that belongs in `seek` instead so that we can still distinguish between "failed to seek to the bookmark you previously set" and "you tried to seek to a non-existent bookmark"? that is, "something failed because of normal situations" vs. "you tried to do something that is guaranteed to fail and should never be done"

I could see that 🤔 The downside is that we'll need to keep the bookmark around -- probably setting it to `null` or so -- to differentiate "invalidated" from "never has been there". Since bookmarks are already scarce, I thought it'd be nice to release them upon invalidation 😄  

Since the user would have to check the return value of `seek()` to find out if the bookmark was valid -- how about instead introducing a `has_bookmark( $bookmark_name )` check for the user to call _before_ `seek()`? (They wouldn't have to do that if they know that they're never touching the HTML in question.)

**Review by adamziel — APPROVED** [review]
This looks great @ockham 🚢 

> Is the safety we're after something that belongs in seek instead

I'm not sure if we need to distinguish between these two situations. If the bookmark is gone, it's gone – the entire code branch relying upon that bookmark will no longer work and needs to perform careful checks. I'm not sure how to exactly use the ability to distinguish between the two failure modes.

**Comment by adamziel** [comment]
> how about instead introducing a has_bookmark( $bookmark_name ) check for the user to call before seek()? (They wouldn't have to do that if they know that they're never touching the HTML in question.)

`has_bookmark` sounds great to me @ockham. Does this PR need anything beyond that?

**Comment by ockham** [comment]
Thank you for reviewing and approving, Adam! (Apologies for not replying any sooner -- I was AFK from Friday until yesterday!)

> > how about instead introducing a has_bookmark( $bookmark_name ) check for the user to call before seek()? (They wouldn't have to do that if they know that they're never touching the HTML in question.)
> 
> `has_bookmark` sounds great to me @ockham. Does this PR need anything beyond that?

No, IMO, that should be all that's needed! 😄 Does that work for you @dmsnell?

**Comment by ockham** [comment]
One thing we need to be mindful of here is that this patch will change the code of the tag processor with regard to what we merged into Core 🤔 While the user is never supposed to encounter any practical consequences from this (since it's currently impossible to make any changes that would actually invalidate bookmakrs), we might want to retain the code in the 6.2 compat layer identical to what is in Core.

So maybe this should go into a different file and/or folder 😬 `lib/compat/wordpress-6.3/` anyone?

**Comment by ockham** [comment]
Rebased on #47933 and modified to add the bookmark invalidation logic to the 6.3 compat layer version of this class.

**Comment by ockham** [comment]
I've added `has_bookmark()` and basic unit test coverage. (Note that we cannot yet test if the invalidation logic work since we simply don't have any method yet that would invalidate bookmarks 😬)

File: lib/compat/wordpress-6.3/html-api/class-gutenberg-html-tag-processor-6-3.php
Comment (dmsnell) [inline lib/compat/wordpress-6.3/html-api/class-gutenberg-html-tag-processor-6-3.php]: I like using variables to assign names to computation, but this one seems a bit redundant. it's almost a literal reading of the computation, so maybe it would make just as much sense to leave this inline?

might also remove a layer of `if` nesting here since there's no use for it. there's _only_ an `A && B` condition here. even if we feel the need to leave a `movl eax, 14 # move 14 into eax`-style comment, we can extract it from the `if` without sacrificing performance.

```php
if ( $bookmark_start_is_after_diff_start && $bookmark->end < $diff->end ) {

}
```

```php
$bookmark_ends_inside_patch = $bookmark->end < $diff->end;

if ( $bookmark_impacted_by_patch && $bookmark_ends_inside_patch ) {

}
```

Comment (dmsnell) [reply lib/compat/wordpress-6.3/html-api/class-gutenberg-html-tag-processor-6-3.php]: played around some more with it and I'm less happy with the rename, particularly since they are _wrong_, including a mismatch between "is_after" and ">=", whereby it may not actually be _after_.

here's a tighter version stewing on relying on the existing names and using `continue 2` to move on to the next bookmark since we know at the point where we're removing the bookmark that we can skip any and all further processing for that bookmark.


not saying I love this version, but I find myself getting lost in the name and purpose of `$bookmark_start_is_after_diff_start`, finding myself wanting instead to move towards the semantic rather than the obvious, something like `$bookmark_shifts` or `$head_shifts` bringing me back to `$update_head`. so this does away with the names altogether; maybe we don't need them.

also I think if we bail before checking whether to release the bookmark we can skip once more some of the zig-zag processing of a bookmark (for release) that we don't need to (because it's unaffected). in other words, instead of repeating the question, "do we need to modify this bookmark?" we can bail immediately and then the rest of the body implies _yes we have to modify it else we would have bailed_.

interested in your thoughts

```php
			foreach ( $this->lexical_updates as $diff ) {
				if ( ! $bookmark->start >= $diff->start && ! $bookmark->end >= $diff->start ) {
					break;
				}

				if ( $bookmark->start >= $diff->start && $bookmark->end < $diff->end ) {
					$this->release_bookmark( $bookmark_name );
					continue 2;
				}

				$delta = strlen( $diff->text ) - ( $diff->end - $diff->start );

				if ( $bookmark->start >= $diff->start ) {
					$head_delta += $delta;
				}

				if ( $bookmark->end >= $diff->end ) {
					$tail_delta += $delta;
				}
			}
```

File: phpunit/html-api/class-gutenberg-html-tag-processor-6-3-test.php
Comment (dmsnell) [inline phpunit/html-api/class-gutenberg-html-tag-processor-6-3-test.php]: not we can verify already in a test if a bookmark exists or not, because `$p->seek( 'place' )` will return `false` if it's unavailable.

```php
$this->assertFalse( $p->seek( 'nonexistent' ) );
$this->assertTrue( $p->seek( 'existent' ) );
// wipe out bookmark with some HTML modifications
$this->assertFalse( $p->seek( 'existent' ) );
```

**Comment by dmsnell** [comment]
Looks very good, and thanks for your patience. I left some questions about the naming, which at first seemed helpful, and then confusing.

Since this is targeting the compat PR can we make sure not to merge it until that one merges? I would rather avoid mixing this _preventative fix_ with that _code shuffle_.

File: lib/compat/wordpress-6.3/html-api/class-gutenberg-html-tag-processor-6-3.php
Comment (ockham) [reply lib/compat/wordpress-6.3/html-api/class-gutenberg-html-tag-processor-6-3.php]: I like it -- thank you for fiddling! 😊 

Sidenote: One thing that's a bit unfortunate with our merge-to-core process is that we don't have the full HTML Processor unit test suite in GB anymore. To verify that this doesn't break any existing tests, I guess I'll copy it to `wordpress-develop` and run unit tests there 😬

Comment (ockham) [reply lib/compat/wordpress-6.3/html-api/class-gutenberg-html-tag-processor-6-3.php]: > To verify that this doesn't break any existing tests, I guess I'll copy it to `wordpress-develop` and run unit tests there 😬 

Tests are passing 👍

Comment (ockham) [reply lib/compat/wordpress-6.3/html-api/class-gutenberg-html-tag-processor-6-3.php]: Applied the suggested change in f171f68adc.

File: phpunit/html-api/class-gutenberg-html-tag-processor-6-3-test.php
Comment (ockham) [reply phpunit/html-api/class-gutenberg-html-tag-processor-6-3-test.php]: Are you suggesting to add a separate test to cover that behavior of `seek`? (Happy to do so 😊 )

Comment (ockham) [reply phpunit/html-api/class-gutenberg-html-tag-processor-6-3-test.php]: Ah, I missed the `// wipe out bookmark with some HTML modifications` part. We don't have any API to do that yet though, do we? 🤔

File: lib/compat/wordpress-6.3/html-api/class-gutenberg-html-tag-processor-6-3.php
Comment (dmsnell) [reply lib/compat/wordpress-6.3/html-api/class-gutenberg-html-tag-processor-6-3.php]: > we don't have the full HTML Processor unit test suite in GB anymore

yep, and this was a big concern of mine too 🤷‍♂️

File: phpunit/html-api/class-gutenberg-html-tag-processor-6-3-test.php
Comment (dmsnell) [reply phpunit/html-api/class-gutenberg-html-tag-processor-6-3-test.php]: > Are you suggesting to add a separate test to cover that behavior of seek?

No; that's a valid thing to talk about, but I was _only_ responding to what I now realize could be a misreading? I thought you had said that it wasn't possible to test _before_ this patch if we have a bookmark or not, and I was suggesting that we _do_, it's just a method that also impacts the processor state.

> We don't have any API to do that yet though, do we? 🤔

That's correct. It's intentionally possible to mess this up though, hopefully in quiet dark corners with blatant warnings to look away. So yes, we could actually write some tests of this behavior before it's necessary, for the purpose of writing tests.

A simpler alternative is simply to assert that a seek to a nonexistent bookmark doesn't exist. I was trying to provide more context in that snippet.

File: lib/compat/wordpress-6.3/html-api/class-gutenberg-html-tag-processor-6-3.php
Comment (dmsnell) [inline lib/compat/wordpress-6.3/html-api/class-gutenberg-html-tag-processor-6-3.php]: we're now dangerously close to some double-negativo here. would the intent of this be clearer?

```php
if ( $bookmark->start < $diff->start && $bookmark->end < $diff->start ) {
	break;
}
```

File: lib/compat/wordpress-6.3/html-api/class-gutenberg-html-tag-processor-6-3.php
Comment (dmsnell) [inline lib/compat/wordpress-6.3/html-api/class-gutenberg-html-tag-processor-6-3.php:1450]: this now reads better to me. I think the symmetry of start/end and the simplicity of bookmark/diff helps us out

File: lib/compat/wordpress-6.3/html-api/class-gutenberg-html-tag-processor-6-3.php
Comment (dmsnell) [inline lib/compat/wordpress-6.3/html-api/class-gutenberg-html-tag-processor-6-3.php]: whether complete sentences or sentence clauses form function summaries.

File: phpunit/html-api/class-gutenberg-html-tag-processor-6-3-test.php
Comment (dmsnell) [inline phpunit/html-api/class-gutenberg-html-tag-processor-6-3-test.php]: we do have a `_bookmark` test file where we're holding bookmarking and seeking behaviors. the main test file is pretty big, and although this is small, it might make sense to keep that separation so that when we merge back into Core it will be easier.

File: lib/compat/wordpress-6.3/html-api/class-gutenberg-html-tag-processor-6-3.php
Comment (ockham) [reply lib/compat/wordpress-6.3/html-api/class-gutenberg-html-tag-processor-6-3.php]: Eh, don't mind me, I'm just copy-pasting [your code](https://github.com/WordPress/gutenberg/pull/47559#discussion_r1103140849) 😆 

But yeah, it did occur to me that it might not be the least unclear thing to do 😏 Happy to change it!

File: lib/compat/wordpress-6.3/html-api/class-gutenberg-html-tag-processor-6-3.php
Comment (ockham) [reply lib/compat/wordpress-6.3/html-api/class-gutenberg-html-tag-processor-6-3.php]: If I haven't read that somewhere as legitimate 🤔 

(Nah, I was probably mixing it up with bool param/return PHPDocs.)

File: lib/compat/wordpress-6.3/html-api/class-gutenberg-html-tag-processor-6-3.php
Comment (adamziel) [reply lib/compat/wordpress-6.3/html-api/class-gutenberg-html-tag-processor-6-3.php]: Wouldn't this suffice?

```php
if ( $bookmark->end < $diff->start ) {
	break;
}
```

I'm assuming that `$bookmark->start < $bookmark->end`

Comment (dmsnell) [reply lib/compat/wordpress-6.3/html-api/class-gutenberg-html-tag-processor-6-3.php]: I think in some ways we're stating that as a constraint, so it should be enough, but I thought it best to leave that safety in as a fallback, since a bug here implies breaking HTML syntax.

also I considered flipping the conditions to check the end first, but then that also ready confusingly, which is what led me to stick with the check I wrote above, a correction apparently to my earlier tangle of code I suggested 😆

File: lib/compat/wordpress-6.3/html-api/class-gutenberg-html-tag-processor-6-3.php
Comment (dmsnell) [reply lib/compat/wordpress-6.3/html-api/class-gutenberg-html-tag-processor-6-3.php]: personally I find it okay for properties, but I look at functions as verbs that _do_ something. the [docs standards](https://developer.wordpress.org/coding-standards/inline-documentation-standards/php/#language) suggest this, but don't state it.

File: lib/compat/wordpress-6.3/html-api/class-gutenberg-html-tag-processor-6-3.php
Comment (ockham) [reply lib/compat/wordpress-6.3/html-api/class-gutenberg-html-tag-processor-6-3.php]: Changed in d91a17fe3e1f2d1426ccf32ea46646e0e839601d.

File: phpunit/html-api/class-gutenberg-html-tag-processor-6-3-test.php
Comment (ockham) [reply phpunit/html-api/class-gutenberg-html-tag-processor-6-3-test.php]: > No; that's a valid thing to talk about, but I was only responding to what I now realize could be a misreading? I thought you had said that it wasn't possible to test before this patch if we have a bookmark or not, and I was suggesting that we do, it's just a method that also impacts the processor state.

Right, thanks for clarifying! 

FWIW, part of the reason for adding `has_bookmark` was to [give the user a way to find out](https://github.com/WordPress/gutenberg/pull/47559#issuecomment-1410220038) without triggering a PHP notice when `WP_DEBUG` is set (as you'd [pointed out](https://github.com/WordPress/gutenberg/pull/47559#pullrequestreview-1275954546)).

I'll add a test for `seek( 'nonexistent' )` though!

Comment (ockham) [reply phpunit/html-api/class-gutenberg-html-tag-processor-6-3-test.php]: > I'll add a test for `seek( 'nonexistent' )` though!

Hmm, this is a bit more involved. Since the PHP notice _is_ being thrown in unit tests, I need the test to "expect" it. 

I just haven't found the right exception type yet 🤔 

```php
$this->expectException( 'PHPUnit_Framework_Error_Notice' ); // Nope, not that one.
```

File: phpunit/html-api/class-gutenberg-html-tag-processor-6-3-test.php
Comment (ockham) [reply phpunit/html-api/class-gutenberg-html-tag-processor-6-3-test.php]: Renamed to include the `-bookmark` suffix (well, infix) in https://github.com/WordPress/gutenberg/pull/47559/commits/f82551d790cbb8d67e804962b1f4585d6b20418b.

(Note that the test file naming convention in GB still diverges from Core's.)

File: lib/compat/wordpress-6.3/html-api/class-gutenberg-html-tag-processor-6-3.php
Comment (ockham) [reply lib/compat/wordpress-6.3/html-api/class-gutenberg-html-tag-processor-6-3.php]: 6d8d6266ba

File: phpunit/html-api/class-gutenberg-html-tag-processor-6-3-test.php
Comment (dmsnell) [reply phpunit/html-api/class-gutenberg-html-tag-processor-6-3-test.php]: > test file naming convention in GB still diverges from Core's

and both diverge from `PHPUnit` :sigh:

**Comment by ockham** [comment]
@dmsnell I think I've addressed all of your feedback; I've only been unable to add test coverage for `seek()` with a non-existent bookmark ([see](https://github.com/WordPress/gutenberg/pull/47559#discussion_r1106367174)).

Good enough? 😬

**Comment by dmsnell** [comment]
> Good enough? 

Yeah I think so, but can we transfer this into a Core ticket as well and put it there?

Pending discussion on how to proceed, what I imagine we might do is end up getting `compat/wordpress-6.3` in as a copy of what's in the long-running branch over at WordPress/WordPress-develop, and once this merges there, we can file a simpler update PR whose description is more like, "pull in updates from 6.3 HTML API"

**Comment by ockham** [comment]
> Pending discussion on how to proceed, what I imagine we might do is end up getting `compat/wordpress-6.3` in as a copy of what's in the long-running branch over at WordPress/WordPress-develop, and once this merges there, we can file a simpler update PR whose description is more like, "pull in updates from 6.3 HTML API"

I replied to this comment on the other PR since it seemed more generally related to the 6.3 compat layer question: https://github.com/WordPress/gutenberg/pull/47933#issuecomment-1432023437 😊

**Comment by ockham** [comment]
Rebased, now that https://github.com/WordPress/gutenberg/pull/47933 has landed 😊 

@adamziel @dmsnell Care to give this another look (and ideally approval)? 😊 🙏

**Comment by gziolo** [comment]
Do we have a WordPress Trac ticket we could reference in the PR? I think it will become an official recommendation soon for PHP changes that we know require backporting to WordPress core.

It would also allow us to include `* @ticket NUMBER` in PHPDoc for newly added tests.

File: lib/compat/wordpress-6.3/html-api/class-gutenberg-html-tag-processor-6-3.php
Comment (adamziel) [inline lib/compat/wordpress-6.3/html-api/class-gutenberg-html-tag-processor-6-3.php:1446]: I still think this would read better as:

```suggestion
				if ( $bookmark->end < $diff->start ) {
```

If `$bookmark->start` is ever after `$bookmark->end`, I'd rather make it fail in some way rather than try to salvage it with a defensive condition.

File: lib/compat/wordpress-6.3/html-api/class-gutenberg-html-tag-processor-6-3.php
Comment (adamziel) [inline lib/compat/wordpress-6.3/html-api/class-gutenberg-html-tag-processor-6-3.php:1457]: ❤️

**Review by adamziel — APPROVED** [review]
I left a nit but it looks great overall. I like how simple it turned!

**Comment by ockham** [comment]
> Do we have a WordPress Trac ticket we could reference in the PR? I think it will become an official recommendation soon for PHP changes that we know require backporting to WordPress core.

We don't have a Trac ticket for this specific change. However, since the general plan is to move development of the HTML (Tag) Processor to `wordpress-develop` (see the conversation in the 6.3 compat layer PR that starts about [here](https://github.com/WordPress/gutenberg/pull/47933#issuecomment-1428620022)), we might file a PR against `wordpress-develop` soon; that would be a good opportunity to also file a Trac ticket, I think.

That ticket might be a bit broader in scope than just this PR, since there's strictly speaking no need for bookmark invalidation as long as there isn't any API in `WP_HTML_Tag_Processor` that could actually cause bookmarks to become invalid. So maybe it'd make more sense to file a ticket to request the ability to make non-attribute lexical changes, which would amount to this PR plus #47068.

> It would also allow us to include `* @ticket NUMBER` in PHPDoc for newly added tests.

That's a good point; but maybe we can do that after the fact? 😊 Especially since the idea is to port changes from `wordpress-develop` back to Gutenberg.

**Comment by gziolo** [comment]
> That ticket might be a bit broader in scope than just this PR, since there's strictly speaking no need for bookmark invalidation as long as there isn't any API in WP_HTML_Tag_Processor that could actually cause bookmarks to become invalid. So maybe it'd make more sense to file a ticket to request the ability to make non-attribute lexical changes, which would amount to this PR plus https://github.com/WordPress/gutenberg/pull/47068.

Yes, one Trac ticket to use as a reference for all work planned in WordPress 6.3 cycle would be perfect. It can cover non-attribute lexical changes for now, but we can extend the scope if that makes sense later.

> > It would also allow us to include * @ticket NUMBER in PHPDoc for newly added tests.
>
> That's a good point; but maybe we can do that after the fact? 😊 Especially since the idea is to port changes from wordpress-develop back to Gutenberg.

I'm not sure I understand the reasoning here. If the ticket blocks merging the PR now, then definitely ignore my comment. We will have to include that annotation in WordPress core anyway, but it's the least important thing today.

File: lib/compat/wordpress-6.3/html-api/class-gutenberg-html-tag-processor-6-3.php
Comment (ockham) [reply lib/compat/wordpress-6.3/html-api/class-gutenberg-html-tag-processor-6-3.php:1446]: While we did inline `$update_head` and `$update_tail` here, we didn't otherwise change the condition -- it existed like that prior to this PR. For that reason, I'd like to keep it as is 😬 I know it shouldn't make a difference outside of any pathological cases, but since this PR has been sitting for a while, I'd like to de-risk it as much as possible.

---

Aside: I kinda like that it's symmetrical to the [condition right below](https://github.com/WordPress/gutenberg/pull/47559/files#diff-345dee9d9f8f95c8f166e2959367d9ed7b67853bb3655586168063defb543119R1450):

```php
if ( $bookmark->start >= $diff->start && $bookmark->end < $diff->end ) {
```

I feel it helps me visualize better that the bookmark is either fully past the diff, or enclosed by it.

---

I'm open to changing this in a follow-up though 😊

**Comment by ockham** [comment]
Thanks @dmsnell @adamziel @gziolo!

It's mergin' time! 💥

**Review by dmsnell — COMMENTED** [review]
this is definitely a good example of why I think developing in Core is a good idea. can we prioritize getting this merged into Core and then remove this code ASAP. right now we have introduced divergent codebases and will need to continue to resolve discrepancies introduced on both sides (something we already did once during the merge of #47933)

that is, remove the unit tests by merging them into Core's suite, then the compat file _probably_ won't need updating (or maybe it _must_ since we'll need to reference a Trac ticket), but _may_ anyway, depending on whether the class in Core received updates as well.

thanks for the work on this.

**Comment by ockham** [comment]
I'll file a PR against `wordpress-develop` later today!

**Comment by ockham** [comment]
> I'll file a PR against `wordpress-develop` later today!

https://github.com/WordPress/wordpress-develop/pull/4116

**Pull request merged**
